### PR TITLE
libgit2: Bump to v1.8.1

### PR DESCRIPTION
**Pull request by bnjmnt4n**
This PR bumps `libgit2` to v1.8.1, in preparation for #1031. I've bumped `libgit2-sys` to `0.17.0+1.8.1`, but haven't touched the `git2` version since I thought that's better left for the maintainers.

I haven't included any new Rust bindings for the new functions (e.g. `git_commit_create_from_stage`). Basically I've just done the bare minimum required to get `git2` working with `libgit2` v1.8.1.

API changes:
- `Remote`:`update_tips` now accepts a `RemoteUpdateFlags` which is an integer consisting of bitfields instead of a boolean
- `FetchOptions`: add `report_unchanged` method which updates the `update_flags` bitflag (similar to above)
- `PushOptions`: add `remote_push_options` method to set remote push options
- `WorktreeAddOptions`: add `checkout_existing` method to set the boolean `checkout_existing` option

## References
- https://github.com/libgit2/libgit2/blob/main/docs/changelog.md#v18
- https://github.com/libgit2/libgit2/blob/main/docs/changelog.md#v181

### CODE REVIEW

**Comment by bnjmnt4n** [comment]
Looks like the current test failures are due to changes made in https://github.com/libgit2/libgit2/pull/6743 👀

**Comment by bnjmnt4n** [comment]
Looks like v1.8.0 is out! I will update the PR to bump the version.

**Comment by bnjmnt4n** [comment]
This should be ready for review. (I'm pretty new to Rust and the `libgit2` / this codebase, please do let me know if there are better ways to do anything.)

The test failures on Windows are related to https://github.com/libgit2/libgit2/pull/6743 as mentioned above—I'm not too sure how to handle these conventionally using Rust.

**Comment by ehuss** [comment]
> The test failures on Windows are related to [libgit2/libgit2#6743](https://github.com/libgit2/libgit2/pull/6743) as mentioned above—I'm not too sure how to handle these conventionally using Rust.

Hm, that's awkward. I don't really understand why TempDir is returning short path. One option is to `canonicalize` both paths before comparing they are equal. Another option is to use [`same-file`](https://docs.rs/same-file/latest/same_file/) to check that the paths are the same. Canonicalize might be easier if it works.

File: libgit2-sys/lib.rs
Comment (bnjmnt4n) [reply libgit2-sys/lib.rs:405]: You're right, I wasn't familiar with C bitfields, I originally thought they were just a convenient way of specifying bit-flags which are handled in a similar way to bitmasks for unsigned ints. It does seem hard to support since their memory layout is compiler dependent. I was thinking since we know the number of bits set, assuming that bits are packed only either left-to-right or right-to-left, we could do something like `update_flag.bits() | update_flag.bits().reverse_bits()` when getting the unsigned int as a workaround? (This admittedly doesn't work if the memory layout is different.)

Comment (extrawurst) [reply libgit2-sys/lib.rs:405]: @ehuss whats the way to interact with c bitfields via FFI?

Comment (bnjmnt4n) [reply libgit2-sys/lib.rs:405]: There is no official support for C bitfields in Rust: https://github.com/rust-lang/rfcs/issues/314.

**Comment by jirutka** [comment]
FYI, this blocks the upgrade of libgit2 to 1.8.0 in Alpine Linux ([!63296](https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/63296)).

**Comment by SS1823** [comment]
Is this PR still blocked?
this may fix https://github.com/actions/checkout/issues/1689

File: libgit2-sys/lib.rs
Comment (ehuss) [reply libgit2-sys/lib.rs:405]: Sorry, I haven't had much time lately. I've been intending to push a change that will use a C wrapper to get/set the fields, I just haven't gotten to it, yet. I don't feel comfortable trying to guess the layout from within Rust. @bnjmnt4n If you have time to give that a shot, that would be helpful. Otherwise, I'll try to look into it soonish.

Comment (kornelski) [reply libgit2-sys/lib.rs:405]: @ehuss @bnjmnt4n The maintainer of libgit2 said he can guarantee the ABI will use least significant bits for the flags: https://github.com/libgit2/libgit2/issues/6800

**Review by kornelski — APPROVED** [review]

File: libgit2-sys/lib.rs
Comment (ehuss) [reply libgit2-sys/lib.rs:405]: Status update: ethomson mentioned in https://github.com/libgit2/libgit2/issues/6800 that they will look to see if there is some workaround for us.

**Comment by jirutka** [comment]
Approved 2 weeks ago, but still not merged… why?

**Comment by kornelski** [comment]
@jirutka libgit 1.8.0 has used C bitfields in its ABI, and that feature is not supported by Rust's FFI and may be compiler-specific.

File: libgit2-sys/lib.rs
Comment (bnjmnt4n) [reply libgit2-sys/lib.rs:405]: I will update this PR when I'm free, since https://github.com/libgit2/libgit2/pull/6806 was merged, but I'm guessing we'll still want to wait for a new release of `libgit2`?

**Comment by bnjmnt4n** [comment]
libgit2 [v1.8.1 was just released](https://github.com/libgit2/libgit2/releases/tag/v1.8.1), and I've updated the PR to use that. Bitfields are no longer used in the new version, so we can avoid any hacks.

**Review by ehuss — APPROVED** [review]
Thanks!

**Pull request merged**
